src/pudl/helpers.py

```diff
@@ -190,52 +190,133 @@ def download_zip_url(
-            lambda x: (
-                af.get_county_fips(state=x[state_col], county=x[county_col])
-                if pd.notnull(x[county_col]) and pd.notnull(x[state_col])
-                else pd.NA
-            ),
-            axis=1,
-        )
-        # force the code columns to be nullable strings - the leading zeros are
-        # important
-        df = df.astype({"county_id_fips": pd.StringDtype()})
-        logger.info(
-            f"Assigned county FIPS codes for "
-            f"{len(df[df.county_id_fips.notnull()]) / len(df):.2%} of records."
+    return df
+
+
+def add_county_fips_id(
+    df: pd.DataFrame, geocodes: pd.DataFrame, county_col: str
+) -> pd.DataFrame:
+    """Add the County FIPS codes to a table with State FIPS codes."""
+    diacretics = {
+        r"ñ": "n",
+        r"'": "",
+        r"ó": "o",
+        r"í": "i",
+        r"á": "a",
+        r"ü": "u",
+        r"é": "e",
+        r"î": "i",
+        r"è": "e",
+        r"à": "a",
+        r"ì": "i",
+        r"å": "a",
+    }
+    abbrevs = {"ft. ": "fort ", "st. ": "saint ", "ste. ": "sainte "}
+
+    county_types = {
+        r" (county|city|city and borough|borough|census area|municipio|municipality|district|parish)$": ""
+    }
+
+    # compile the counties.
+    counties = geocodes.loc[
+        geocodes["fips_level"] == "050",
+        ["area_name", "state_id_fips", "county_id_fips"],
+    ].assign(  # squish the state and county fips together to get the full code
+        county_id_fips=lambda x: x.state_id_fips + x.county_id_fips
+    )
+    # compile all of the many possible ways these county columns could show up.
+    counties = pd.concat(
+        [
+            counties.assign(
+                county_tmp=_clean_area_name_col(counties["area_name"], replace_dict)
+            )
+            for replace_dict in [
+                {},
+                diacretics | abbrevs,
+                county_types,
+                diacretics | abbrevs | county_types,
+                county_types | {"'s": "s"},
+            ]
+        ]
+    ).drop_duplicates()
+
+    # see assertion note below. we keep the city records because all instances
+    # of the city as county reported we've seen always note lone city as county
+    # name instead of lone county
+    city_county_dupe_mask = counties.duplicated(
+        ["state_id_fips", "county_tmp"], keep=False
+    ) & ~(counties.area_name.str.lower().str.contains("city"))
+
+    if len(city_county_dupes := counties[city_county_dupe_mask]) > 10:
+        raise AssertionError(
+            "We expect there to be 10 county records that have the same name and state "
+            "but have different county fips codes. This is due to pair counties that have "
+            "a city jurisdiction and a corresponding county jurisdiction (ex. Baltimore "
+            "County and Baltimore City which is also a county). We found "
+            f"{len(city_county_dupes)}:\n{city_county_dupes}"
```

Comment:
lol jk i can't choose these cities across the board. there are only a handful of them but i will need to specify which to choose.